common/components/org.wso2.carbon.data.agents.log4j/src/main/java/org/wso2/carbon/data/agents/log4j/appender/LogEventAppender.java

```diff
@@ -0,0 +1,411 @@
+                            getMaxTolerableConsecutiveFailure() + ". Purging log event array. Some logs will be lost.");
+                    loggingEvents.clear();
+                    numOfConsecutiveFailures = 0;
+                }
+            }
+        }
+
+        private void publishLogEvent(TenantAwareLoggingEvent event)
+                throws ParseException, MalformedURLException, AuthenticationException, TransportException,
+                MalformedStreamDefinitionException, StreamDefinitionException,
+                DifferentStreamDefinitionAlreadyDefinedException, ExecutionException {
+            String streamId = "";
+            String streamName = "";
+            String tenantId = event.getTenantId();
+            if (tenantId.equals(String.valueOf(MultitenantConstants.INVALID_TENANT_ID)) || tenantId
+                    .equals(String.valueOf(MultitenantConstants.SUPER_TENANT_ID))) {
+                tenantId = "0";
+            }
+            String serverKey = getCurrentServerName();
+            String currDateStr = getCurrentDate();
+            try {
+                Pattern pattern = Pattern.compile("(tcp):\\/\\/([a-zA-Z0-9]+):([0-9]+)");
+                Matcher matcher = pattern.matcher(url);
+                if (matcher.find() && authURLSet == null) {
+                    authURLSet = "ssl://" + matcher.group(2).toString() + ":" + (
+                            Integer.parseInt(matcher.group(3).toString()) + 100);
+                }
+                dataPublisher = new DataPublisher("Thrift", url, authURLSet, userName, password);
+            } catch (DataEndpointAgentConfigurationException e) {
+                log.error(
+                        "Error in initializing binary data-publisher to send requests to global throttling engine " + e
+                                .getMessage(), e);
+            } catch (DataEndpointException e) {
+                log.error(
+                        "Error in initializing binary data-publisher to send requests to global throttling engine " + e
+                                .getMessage(), e);
+            } catch (DataEndpointConfigurationException e) {
+                log.error(
+                        "Error in initializing binary data-publisher to send requests to global throttling engine " + e
+                                .getMessage(), e);
+            } catch (DataEndpointAuthenticationException e) {
+                log.error(
+                        "Error in initializing binary data-publisher to send requests to global throttling engine " + e
+                                .getMessage(), e);
+            } catch (TransportException e) {
+                log.error(
+                        "Error in initializing binary data-publisher to send requests to global throttling engine " + e
+                                .getMessage(), e);
+            }
```

Comment:
As per the above approach when publishing the log event it will check whether the publisher is null or not , if it is null then it tries to initialize it again , after that it will check the publisher null state again and if null will throw an exception which has  content saying  publisher not initialed for the given url.

This will print System.error for each and every log event saying the publisher is not initializing for the url not a NPE.

I understand that this is not the best approach to handle the publisher initialization, since the all other task are depend on the data publisher. But we cant initialize the data publisher in the appender activateOptions(), because this data publisher needs some system property to set for the initialization process (ex : "javax.net.ssl.trustStore").  When the carbon logging invoke this appender activateOptions() method those properties are not setup to the system.

As per the new approach, we try to initialize data publisher for the first event checking nullity of the publisher object. If publisher is null log event processing will be stopped until user provide correct publisher configuration and restart the server. This will prevent multiple log messages.

Fixes for the new approach are in : https://github.com/wso2/shared-analytics/pull/131/files#diff-d45f010b040aded5b5711f7ddab28e59R216